src/effects/ending_decals.c

```diff
@@ -1,14 +1,97 @@
 #include "common.h"
 #include "effects_internal.h"
 
+extern Gfx D_09001E40;
+extern Gfx* D_E00685B0[];
+extern Gfx* D_E00685BC[];
+extern Gfx* D_E00685F4[];
+
+void ending_decals_init(EffectInstance* effect);
+void ending_decals_update(EffectInstance* effect);
+void ending_decals_render(EffectInstance* effect);
 void ending_decals_appendGfx(void* effect);
 
-INCLUDE_ASM(s32, "effects/ending_decals", ending_decals_main);
+void ending_decals_main(s32 arg0, f32 arg1, f32 arg2, f32 arg3, f32 arg4, EffectInstance** outEffect) {
+    EffectBlueprint bp;
+    EffectInstance* effect;
+    EndingDecalsFXData* data;
+    s32 numParts = 1;
+
+    bp.init = ending_decals_init;
+    bp.update = ending_decals_update;
+    bp.renderWorld = ending_decals_render;
+    bp.unk_00 = 0;
+    bp.unk_14 = NULL;
+    bp.effectID = EFFECT_ENDING_DECALS;
+
+    effect = shim_create_effect_instance(&bp);
+    effect->numParts = numParts;
+    data = effect->data.endingDecals = shim_general_heap_malloc(numParts * sizeof(*data));
+    ASSERT(effect->data.endingDecals != NULL);
+
+    data->unk_00 = arg0;
+    data->pos.x = arg1;
+    data->pos.y = arg2;
+    data->pos.z = arg3;
+    data->unk_18 = 255;
+    data->unk_14 = arg4;
+    data->unk_10 = arg4;
+    data->unk_1C = 100;
+    data->unk_20 = 0;
+
+    if (arg0 >= 0) {
+        if (arg0 < 2) {
+            data->unk_24 = 0;
+            data->unk_25 = 0;
+            data->unk_26 = 0;
+            data->unk_27 = 255;
+            data->unk_28 = 255;
+            data->unk_29 = 255;
+        } else {
+            goto block_6;
```

Comment:
Couldn't remove this `goto`. Feel free to give it a try: https://decomp.me/scratch/1fp80